File: src/main/java/org/apache/fineract/client/services/ClientTransactionApi.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.fineract.client.services;

import org.apache.fineract.client.CollectionFormats.*;

import io.reactivex.Observable;
import io.reactivex.Completable;
import retrofit2.http.*;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.MultipartBody;

import org.apache.fineract.client.models.GetClientsClientIdTransactionsResponse;
import org.apache.fineract.client.models.GetClientsClientIdTransactionsTransactionIdResponse;
import org.apache.fineract.client.models.PostClientsClientIdTransactionsTransactionIdResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public interface ClientTransactionApi {
  /**
   * List Client Transactions
   * The list capability of client transaction can support pagination.  Example Requests:  clients/189/transactions  clients/189/transactions?offset&#x3D;10&amp;limit&#x3D;50
   * @param clientId clientId (required)
   * @param offset offset (optional)
   * @param limit limit (optional)
   * @return Observable&lt;GetClientsClientIdTransactionsResponse&gt;
   */
  @GET("clients/{clientId}/transactions")
  Observable<GetClientsClientIdTransactionsResponse> retrieveAllClientTransactions(
    @retrofit2.http.Path("clientId") Long clientId, @retrofit2.http.Query("offset") Integer offset, @retrofit2.http.Query("limit") Integer limit
  );

  /**
   * Retrieve a Client Transaction
   * Example Requests: clients/1/transactions/1   clients/1/transactions/1?fields&#x3D;id,officeName
   * @param clientId clientId (required)
   * @param transactionId transactionId (required)
   * @return Observable&lt;GetClientsClientIdTransactionsTransactionIdResponse&gt;
   */
  @GET("clients/{clientId}/transactions/{transactionId}")
  Observable<GetClientsClientIdTransactionsTransactionIdResponse> retrieveClientTransaction(
    @retrofit2.http.Path("clientId") Long clientId, @retrofit2.http.Path("transactionId") Long transactionId
  );

  /**
   * Undo a Client Transaction
   * Undoes a Client Transaction
   * @param clientId clientId (required)
   * @param transactionId transactionId (required)
   * @param command command (optional)
   * @return Observable&lt;PostClientsClientIdTransactionsTransactionIdResponse&gt;
   */
  @POST("clients/{clientId}/transactions/{transactionId}")
  Observable<PostClientsClientIdTransactionsTransactionIdResponse> undoClientTransaction(
    @retrofit2.http.Path("clientId") Long clientId, @retrofit2.http.Path("transactionId") Long transactionId, @retrofit2.http.Query("command") String command
  );

}

File: src/main/kotlin/org/openapitools/client/apis/ClientTransactionApi.kt
package org.openapitools.client.apis

import org.openapitools.client.models.GetClientsClientIdTransactionsResponse
import org.openapitools.client.models.GetClientsClientIdTransactionsTransactionIdResponse
import org.openapitools.client.models.PostClientsClientIdTransactionsTransactionIdResponse
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface ClientTransactionApi {
    /**
     * List Client Transactions
     * The list capability of client transaction can support pagination.  Example Requests:  clients/189/transactions  clients/189/transactions?offset&#x3D;10&amp;limit&#x3D;50
     * Responses:
     *  - 200: OK
     *
     * @param clientId clientId
     * @param offset offset (optional)
     * @param limit limit (optional)
     * @return [GetClientsClientIdTransactionsResponse]
     */
    @GET("v1/clients/{clientId}/transactions")
    suspend fun retrieveAllClientTransactions(
        @Path("clientId") clientId: Long,
        @Query("offset") offset: Int? = null,
        @Query("limit") limit: Int? = null
    ): GetClientsClientIdTransactionsResponse

    /**
     * List Client Transactions
     * The list capability of client transaction can support pagination.  Example Requests:  clients/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854/transactions  clients/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854/transactions?offset&#x3D;10&amp;limit&#x3D;50
     * Responses:
     *  - 200: OK
     *
     * @param clientExternalId clientExternalId
     * @param offset offset (optional)
     * @param limit limit (optional)
     * @return [GetClientsClientIdTransactionsResponse]
     */
    @GET("v1/clients/external-id/{clientExternalId}/transactions")
    suspend fun retrieveAllClientTransactions1(
        @Path("clientExternalId") clientExternalId: String,
        @Query("offset") offset: Int? = null,
        @Query("limit") limit: Int? = null
    ): GetClientsClientIdTransactionsResponse

    /**
     * Retrieve a Client Transaction
     * Example Requests: clients/1/transactions/1   clients/1/transactions/1?fields&#x3D;id,officeName
     * Responses:
     *  - 200: OK
     *
     * @param clientId clientId
     * @param transactionId transactionId
     * @return [GetClientsClientIdTransactionsTransactionIdResponse]
     */
    @GET("v1/clients/{clientId}/transactions/{transactionId}")
    suspend fun retrieveClientTransaction(
        @Path("clientId") clientId: Long,
        @Path("transactionId") transactionId: Long
    ): GetClientsClientIdTransactionsTransactionIdResponse

    /**
     * Retrieve a Client Transaction
     * Example Requests: clients/1/transactions/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854   clients/1/transactions/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854?fields&#x3D;id,officeName
     * Responses:
     *  - 200: OK
     *
     * @param clientId clientId
     * @param transactionExternalId transactionExternalId
     * @return [GetClientsClientIdTransactionsTransactionIdResponse]
     */
    @GET("v1/clients/{clientId}/transactions/external-id/{transactionExternalId}")
    suspend fun retrieveClientTransaction1(
        @Path("clientId") clientId: Long,
        @Path("transactionExternalId") transactionExternalId: String
    ): GetClientsClientIdTransactionsTransactionIdResponse

    /**
     * Retrieve a Client Transaction
     * Example Requests: clients/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854/transactions/1   clients/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854/transactions/1?fields&#x3D;id,officeName
     * Responses:
     *  - 200: OK
     *
     * @param clientExternalId clientExternalId
     * @param transactionId transactionId
     * @return [GetClientsClientIdTransactionsTransactionIdResponse]
     */
    @GET("v1/clients/external-id/{clientExternalId}/transactions/{transactionId}")
    suspend fun retrieveClientTransaction2(
        @Path("clientExternalId") clientExternalId: String,
        @Path("transactionId") transactionId: Long
    ): GetClientsClientIdTransactionsTransactionIdResponse

    /**
     * Retrieve a Client Transaction
     * Example Requests: clients/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854/transactions/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854   clients/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854/transactions/external-id/7dd80a7c-ycba-a446-t378-91eb6f53e854?fields&#x3D;id,officeName
     * Responses:
     *  - 200: OK
     *
     * @param clientExternalId clientExternalId
     * @param transactionExternalId transactionExternalId
     * @return [GetClientsClientIdTransactionsTransactionIdResponse]
     */
    @GET("v1/clients/external-id/{clientExternalId}/transactions/external-id/{transactionExternalId}")
    suspend fun retrieveClientTransaction3(
        @Path("clientExternalId") clientExternalId: String,
        @Path("transactionExternalId") transactionExternalId: String
    ): GetClientsClientIdTransactionsTransactionIdResponse

    /**
     * Undo a Client Transaction
     * Undoes a Client Transaction
     * Responses:
     *  - 200: OK
     *
     * @param clientId clientId
     * @param transactionId transactionId
     * @param command command (optional)
     * @return [PostClientsClientIdTransactionsTransactionIdResponse]
     */
    @POST("v1/clients/{clientId}/transactions/{transactionId}")
    suspend fun undoClientTransaction(
        @Path("clientId") clientId: Long,
        @Path("transactionId") transactionId: Long,
        @Query("command") command: String? = null
    ): PostClientsClientIdTransactionsTransactionIdResponse

    /**
     * Undo a Client Transaction
     * Undoes a Client Transaction
     * Responses:
     *  - 200: OK
     *
     * @param clientId clientId
     * @param transactionExternalId transactionExternalId
     * @param command command (optional)
     * @return [PostClientsClientIdTransactionsTransactionIdResponse]
     */
    @POST("v1/clients/{clientId}/transactions/external-id/{transactionExternalId}")
    suspend fun undoClientTransaction1(
        @Path("clientId") clientId: Long,
        @Path("transactionExternalId") transactionExternalId: String,
        @Query("command") command: String? = null
    ): PostClientsClientIdTransactionsTransactionIdResponse

    /**
     * Undo a Client Transaction
     * Undoes a Client Transaction
     * Responses:
     *  - 200: OK
     *
     * @param clientExternalId clientExternalId
     * @param transactionId transactionId
     * @param command command (optional)
     * @return [PostClientsClientIdTransactionsTransactionIdResponse]
     */
    @POST("v1/clients/external-id/{clientExternalId}/transactions/{transactionId}")
    suspend fun undoClientTransaction2(
        @Path("clientExternalId") clientExternalId: String,
        @Path("transactionId") transactionId: Long,
        @Query("command") command: String? = null
    ): PostClientsClientIdTransactionsTransactionIdResponse

    /**
     * Undo a Client Transaction
     * Undoes a Client Transaction
     * Responses:
     *  - 200: OK
     *
     * @param clientExternalId clientExternalId
     * @param transactionExternalId transactionExternalId
     * @param command command (optional)
     * @return [PostClientsClientIdTransactionsTransactionIdResponse]
     */
    @POST("v1/clients/external-id/{clientExternalId}/transactions/external-id/{transactionExternalId}")
    suspend fun undoClientTransaction3(
        @Path("clientExternalId") clientExternalId: String,
        @Path("transactionExternalId") transactionExternalId: String,
        @Query("command") command: String? = null
    ): PostClientsClientIdTransactionsTransactionIdResponse

}
